pyomo/gdp/plugins/transform_current_disjunctive_logic.py

```diff
@@ -0,0 +1,213 @@
+                blocks = t.values() if t.is_indexed() else (t,)
+                for block in blocks:
+                    self._transform_block(block, reverse_token, disjunction_transform)
+            elif t.ctype is Disjunction:
+                disjunctions = t.values() if t.is_indexed() else (t,)
+                for disj in disjunctions:
+                    disjunction_transform(disj, reverse_token)
+        if config.reverse is None:
+            return reverse_token
+
+    def _transform_block(self, block, reverse_token, disjunction_transform):
+        for disjunction in block.component_data_objects(
+            Disjunction,
+            active=True,
+            descend_into=(Block, Disjunct),
+            sort=SortComponents.deterministic,
+        ):
+            disjunction_transform(disjunction, reverse_token)
+
+    def _transform_disjunction(self, disjunction, reverse_token):
+        no_val = set()
+        true_val = set()
+        false_val = set()
+        num_disjuncts = 0
+        for disjunct in disjunction.disjuncts:
+            if not disjunct.active:
+                continue
+            num_disjuncts += 1
+            ind_var = disjunct.indicator_var
+            if ind_var.value is None:
+                no_val.add(disjunct)
+            elif ind_var.value:
+                true_val.add(disjunct)
+            else:
+                false_val.add(disjunct)
+        if disjunction.xor:
+            if len(true_val) > 1 or len(false_val) == num_disjuncts:
+                raise InfeasibleConstraintException(
+                    "Exactly-one constraint for Disjunction "
+                    "'%s' is violated. The following Disjuncts "
+                    "are selected: %s"
+                    % (disjunction.name, ', '.join([d.name for d in true_val]))
+                )
+            elif len(true_val) == 1:
+                # We can fix everything
+                self._reclassify_disjuncts(
+                    true_val, no_val.union(false_val), reverse_token['_disjuncts']
+                )
+                # This disjunction is fully transformed
+                reverse_token['_disjunctions'].add(disjunction)
+                disjunction.deactivate()
+            elif len(false_val) == num_disjuncts - 1:
+                # We can fix everything. Since we didn't hit the case above, we
+                # know that the non-False value is a None.
+                self._reclassify_disjuncts(
+                    no_val, false_val, reverse_token['_disjuncts']
+                )
+                reverse_token['_disjunctions'].add(disjunction)
+                disjunction.deactivate()
+        # It's only an 'at-least' not an 'exactly'...
+        elif len(false_val) == num_disjuncts:
+            raise InfeasibleConstraintException(
+                "Atleast-one constraint for Disjunction "
+                "'%s' is violated. That is, all the Disjunct "
+                "indicator_vars are 'False'." % disjunction.name
```

Comment:
Should this exception be copied above in case all disjunct are False?

Reply:
Ah, good point, I can check for this one before I even need to know about exactly-one or at-least. That will probably make a cleaner error message.